src/user-session-helper/sdi-user-session-helper.c

```diff
@@ -0,0 +1,255 @@
+
+static guint dbus_check_id = 0;
+static guint dbus_timeout_id = 0;
+static gchar *dbus_names[] = {"org.freedesktop.portal.Desktop",
+                              "*org.freedesktop.impl.portal.desktop.", NULL};
+
+static gboolean sdi_session_is_desktop(const gchar *object_path) {
+  g_autoptr(OrgFreedesktopLogin1Session) session = NULL;
+  g_autoptr(GVariant) user = NULL;
+  // these values belongs to the session proxy, so they must not be freed
+  GVariant *user_data = NULL;
+  const gchar *session_type = NULL;
+
+  session = org_freedesktop_login1_session_proxy_new_for_bus_sync(
+      G_BUS_TYPE_SYSTEM, G_DBUS_PROXY_FLAGS_NONE, "org.freedesktop.login1",
+      object_path, NULL, NULL);
+  user_data = org_freedesktop_login1_session_get_user(session);
+  if (user_data == NULL) {
+    g_warning("Failed to read the session user data. Forcing a reload.");
+    /* if we can't read the data, we can't know whether we are in a desktop
+     * session or in a text one, so we will assume that we are in a session
+     * desktop to force a reload.
+     */
+    return TRUE;
+  }
+  user = g_variant_get_child_value(user_data, 0);
+  if (user == NULL) {
+    g_warning("Failed to read the session user.");
+    return FALSE;
+  }
+  if (getuid() != g_variant_get_uint32(user)) {
+    // the new session isn't for our user
+    return FALSE;
+  }
+  session_type = org_freedesktop_login1_session_get_type_(session);
+  if (session_type == NULL) {
+    g_warning("Failed to read the session type");
+    return FALSE;
+  }
+  if (g_str_equal("x11", session_type) ||
+      g_str_equal("wayland", session_type) ||
+      g_str_equal("mir", session_type)) {
+    // this is a graphical session
+    return TRUE;
+  }
+  return FALSE;
+}
+
+static gboolean sdi_check_graphical_sessions(Login1Manager *login_manager) {
+  GVariant *sessions = NULL;
+  gboolean got_session_list;
+
+  got_session_list = login1_manager_call_list_sessions_sync(
+      login_manager, &sessions, NULL, NULL);
+
+  if (!got_session_list) {
+    g_warning("Failed to get session list (check that login-session-observe "
+              "interface is connected). Forcing a reload.");
+    return false;
+  }
+
+  /* check if there is already a graphical session opened for us, in which
+   * case we must just exit and let systemd to relaunch us, because it means
+   * that we run too early and the desktop wasn't still ready.
+   */
+  for (int i = 0; i < g_variant_n_children(sessions); i++) {
+    g_autoptr(GVariant) session = g_variant_get_child_value(sessions, i);
+    if (session == NULL) {
+      continue;
+    }
+    g_autoptr(GVariant) session_object_variant =
+        g_variant_get_child_value(session, 4);
+    const gchar *session_object =
+        g_variant_get_string(session_object_variant, NULL);
+    g_debug("Checking session %s...", session_object);
+    if (sdi_session_is_desktop(session_object)) {
+      g_debug("We are already in a desktop session!");
+      return true;
+    }
+  }
```

Comment:
Also here, can just use `g_variant_get` and then iterate over a GVariantIter instead of manual (and partially unchecked) unpacking.